pkg/plugin/plugin.go

```diff
@@ -36,15 +38,67 @@ import (
 	"github.com/projectcalico/libcalico-go/lib/options"
 	"github.com/sirupsen/logrus"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
 )
 
+const testConnectionTimeout = 2 * time.Second
+
 func init() {
 	// This ensures that main runs only on main thread (thread group leader).
 	// since namespace ops (unshare, setns) are done for a single thread, we
 	// must ensure that the goroutine does not jump from OS thread to thread
 	runtime.LockOSThread()
 }
 
+func testConnection() error {
+	// Unmarshal the network config
+	conf := types.NetConf{}
+	data, err := ioutil.ReadAll(os.Stdin)
+	if err != nil {
+		return errors.New("failed to read from stdin")
+	}
+	if err := json.Unmarshal(data, &conf); err != nil {
+		return fmt.Errorf("failed to load netconf: %v", err)
+	}
+
+	// Create a new client.
+	calicoClient, err := utils.CreateClient(conf)
+	if err != nil {
+		return err
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), testConnectionTimeout)
+	defer cancel()
+	ci, err := calicoClient.ClusterInformation().Get(ctx, "default", options.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("error getting ClusterInformation: %v", err)
+	}
+	if !*ci.Spec.DatastoreReady {
+		logrus.Info("Upgrade may be in progress, ready flag is not set")
+		return fmt.Errorf("Calico is currently not ready to process requests")
+	}
+
+	// If we have a kubeconfig, test connection to the APIServer
+	if conf.Kubernetes.Kubeconfig != "" {
```

Comment:
If we're using KDD mode, then we'll have already tested this above. Maybe worth gating this check on etcd mode only? Then again, it doesn't _hurt_ to have two checks 🤔